File: app/src/main/java/com/example/shellshock/TopUpActivity.java
package com.example.shellshock;

import android.app.PendingIntent;
import android.content.Intent;
import android.nfc.NfcAdapter;
import android.nfc.Tag;
import android.nfc.tech.IsoDep;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.text.InputType;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.google.android.material.textfield.TextInputEditText;

import java.io.IOException;
import java.util.concurrent.CompletableFuture;

public class TopUpActivity extends AppCompatActivity {

    private static final String TAG = "TopUpActivity";
    private TextInputEditText proofTokenEditText;
    private Button topUpSubmitButton;
    private TextView statusTextView;
    private AlertDialog nfcDialog;
    private NfcAdapter nfcAdapter;
    private SatocashNfcClient satocashClient;
    private SatocashWallet satocashWallet;
    private String pendingProofToken;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_top_up);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setTitle("Top Up Wallet");
        }

        proofTokenEditText = findViewById(R.id.top_up_amount_edit_text);
        topUpSubmitButton = findViewById(R.id.top_up_submit_button);
        statusTextView = findViewById(R.id.statusTextView);

        nfcAdapter = NfcAdapter.getDefaultAdapter(this);
        if (nfcAdapter == null) {
            logStatus("NFC is not available on this device. Cannot flash proofs.");
            topUpSubmitButton.setEnabled(false);
        } else {
            logStatus("NFC available. Enter token or place card to flash directly.");
        }

        topUpSubmitButton.setOnClickListener(v -> {
            String proofToken = proofTokenEditText.getText().toString();
            if (!proofToken.isEmpty()) {
                pendingProofToken = proofToken;
                logStatus("Token set. Ready to flash to card: " + proofToken.substring(0, Math.min(proofToken.length(), 30)) + "...");
                Toast.makeText(this, "Token set. Now tap a card to import proofs.", Toast.LENGTH_LONG).show();
                showNfcDialog();
            } else {
                logStatus("Please enter a Cashu proof token first.");
                Toast.makeText(this, "Please enter a Cashu proof token", Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }

    private void logStatus(String message) {
        Log.d(TAG, message);
        mainHandler.post(() -> {
            statusTextView.append(message + "\n");
            // Scroll to the bottom
            int scrollAmount = statusTextView.getLayout().getLineTop(statusTextView.getLineCount()) - statusTextView.getHeight();
            if (scrollAmount > 0) {
                statusTextView.scrollTo(0, scrollAmount);
            } else {
                statusTextView.scrollTo(0, 0);
            }
        });
    }

    private void showNfcDialog() {
        mainHandler.post(() -> {
            AlertDialog.Builder builder = new AlertDialog.Builder(this, R.style.Theme_ShellShock);
            LayoutInflater inflater = this.getLayoutInflater();
            View dialogView = inflater.inflate(R.layout.dialog_nfc_modern, null);
            builder.setView(dialogView);

            TextView nfcAmountDisplay = dialogView.findViewById(R.id.nfc_amount_display);
            nfcAmountDisplay.setText("Ready to import proofs");

            builder.setCancelable(true);
            nfcDialog = builder.create();
            nfcDialog.show();
        });
    }

    private void showPinDialog(PinDialogCallback callback) {
        mainHandler.post(() -> {
            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setTitle("Enter PIN");

            // Create layout for PIN input
            LinearLayout layout = new LinearLayout(this);
            layout.setOrientation(LinearLayout.VERTICAL);
            int padding = (int) (50 * getResources().getDisplayMetrics().density);
            int paddingVertical = (int) (20 * getResources().getDisplayMetrics().density);
            layout.setPadding(padding, paddingVertical, padding, paddingVertical);

            // Add PIN input field
            EditText input = new EditText(this);
            input.setInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_VARIATION_PASSWORD);
            input.setHint("PIN");
            layout.addView(input);

            // Add keypad layout
            LinearLayout keypadLayout = new LinearLayout(this);
            keypadLayout.setOrientation(LinearLayout.VERTICAL);
            LinearLayout.LayoutParams keypadParams = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
            );
            keypadLayout.setLayoutParams(keypadParams);

            String[][] buttons = {
                    {"1", "2", "3"},
                    {"4", "5", "6"},
                    {"7", "8", "9"},
                    {"", "0", "DEL"}
            };

            for (String[] row : buttons) {
                LinearLayout rowLayout = new LinearLayout(this);
                rowLayout.setOrientation(LinearLayout.HORIZONTAL);
                LinearLayout.LayoutParams rowParams = new LinearLayout.LayoutParams(
                        LinearLayout.LayoutParams.MATCH_PARENT,
                        LinearLayout.LayoutParams.WRAP_CONTENT
                );
                rowParams.weight = 1.0f;
                rowLayout.setLayoutParams(rowParams);

                for (String text : row) {
                    Button button = new Button(this);
                    button.setText(text);
                    LinearLayout.LayoutParams buttonParams = new LinearLayout.LayoutParams(
                            0,
                            LinearLayout.LayoutParams.WRAP_CONTENT
                    );
                    buttonParams.weight = 1.0f;
                    button.setLayoutParams(buttonParams);

                    button.setOnClickListener(v -> {
                        if ("DEL".equals(text)) {
                            if (input.length() > 0) {
                                input.getText().delete(input.length() - 1, input.length());
                            }
                        } else if (!text.isEmpty()) {
                            input.append(text);
                        }
                    });
                    rowLayout.addView(button);
                }
                keypadLayout.addView(rowLayout);
            }

            // Add button layout
            LinearLayout buttonLayout = new LinearLayout(this);
            buttonLayout.setOrientation(LinearLayout.HORIZONTAL);
            LinearLayout.LayoutParams buttonLayoutParams = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
            );
            buttonLayoutParams.topMargin = (int) (20 * getResources().getDisplayMetrics().density);
            buttonLayout.setLayoutParams(buttonLayoutParams);

            // Cancel button
            Button cancelButton = new Button(this);
            cancelButton.setText("Cancel");
            LinearLayout.LayoutParams cancelParams = new LinearLayout.LayoutParams(
                    0,
                    LinearLayout.LayoutParams.WRAP_CONTENT,
                    1.0f
            );
            cancelParams.rightMargin = (int) (8 * getResources().getDisplayMetrics().density);
            cancelButton.setLayoutParams(cancelParams);

            // OK button
            Button okButton = new Button(this);
            okButton.setText("OK");
            LinearLayout.LayoutParams okParams = new LinearLayout.LayoutParams(
                    0,
                    LinearLayout.LayoutParams.WRAP_CONTENT,
                    1.0f
            );
            okParams.leftMargin = (int) (8 * getResources().getDisplayMetrics().density);
            okButton.setLayoutParams(okParams);

            buttonLayout.addView(cancelButton);
            buttonLayout.addView(okButton);

            layout.addView(keypadLayout);
            layout.addView(buttonLayout);
            builder.setView(layout);

            AlertDialog dialog = builder.create();

            cancelButton.setOnClickListener(v -> {
                dialog.cancel();
                callback.onPin(null);
            });

            okButton.setOnClickListener(v -> {
                String pin = input.getText().toString();
                dialog.dismiss();
                callback.onPin(pin);
            });

            dialog.setOnCancelListener(dialogInterface -> callback.onPin(null));

            dialog.show();
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (nfcAdapter != null) {
            PendingIntent pendingIntent = PendingIntent.getActivity(
                    this, 0, new Intent(this, getClass())
                            .addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP),
                    PendingIntent.FLAG_MUTABLE);
            String[][] techList = new String[][]{new String[]{IsoDep.class.getName()}};
            nfcAdapter.enableForegroundDispatch(this, pendingIntent, null, techList);
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (nfcAdapter != null) {
            nfcAdapter.disableForegroundDispatch(this);
        }
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        if (NfcAdapter.ACTION_TECH_DISCOVERED.equals(intent.getAction())) {
            Tag tag = intent.getParcelableExtra(NfcAdapter.EXTRA_TAG);
            if (tag != null) {
                handleNfcImport(tag);
            }
        }
    }

    private void handleNfcImport(Tag tag) {
        if (pendingProofToken == null || pendingProofToken.isEmpty()) {
            logStatus("No proof token set to import. Please enter it first.");
            Toast.makeText(this, "No token to import. Please enter one first.", Toast.LENGTH_LONG).show();
            return;
        }

        logStatus("NFC Tag discovered. Attempting to import proofs...");

        new Thread(() -> {
            try {
                satocashClient = new SatocashNfcClient(tag);
                satocashClient.connect();
                satocashWallet = new SatocashWallet(satocashClient);

                logStatus("Selecting Satocash Applet...");
                satocashClient.selectApplet(SatocashNfcClient.SATOCASH_AID);
                logStatus("Satocash Applet found and selected!");

                logStatus("Initializing Secure Channel...");
                satocashClient.initSecureChannel();
                logStatus("Secure Channel Initialized!");

                // Create a CompletableFuture for the PIN dialog result
                CompletableFuture<String> pinFuture = new CompletableFuture<>();
                showPinDialog(pin -> pinFuture.complete(pin));
                String pin = pinFuture.join();

                if (pin != null) {
                    try {
                        CompletableFuture<Boolean> authFuture = satocashWallet.authenticatePIN(pin);
                        boolean authenticated = authFuture.join();
                        
                        if (authenticated) {
                            logStatus("PIN Verified. Importing proofs...");

                            try {
                                CompletableFuture<Integer> importFuture = satocashWallet.importProofsFromToken(pendingProofToken);
                                int importedCount = importFuture.join();
                                logStatus("Successfully imported " + importedCount + " proofs to card!");

                                mainHandler.post(() -> {
                                    if (nfcDialog != null && nfcDialog.isShowing()) {
                                        nfcDialog.dismiss();
                                    }
                                    Toast.makeText(TopUpActivity.this,
                                            "Imported " + importedCount + " proofs!", Toast.LENGTH_LONG).show();
                                    pendingProofToken = "";
                                });
                            } catch (Exception e) {
                                String message = "Import failed: " + e.getMessage();
                                logStatus(message);
                                Log.e(TAG, message, e);
                            }
                        } else {
                            String message = "PIN Verification Failed";
                            logStatus(message);
                            Log.e(TAG, message);
                        }

                    } catch (RuntimeException e) {
                        Throwable cause = e.getCause();
                        if (cause instanceof SatocashNfcClient.SatocashException) {
                            SatocashNfcClient.SatocashException satocashEx = (SatocashNfcClient.SatocashException) cause;
                            String message = String.format("PIN Verification Failed: %s (SW: 0x%04X)",
                                    satocashEx.getMessage(), satocashEx.getSw());
                            logStatus(message);
                            Log.e(TAG, message);
                        } else {
                            String message = "SatocashWallet Failed: " + e.getMessage();
                            logStatus(message);
                            Log.e(TAG, message, e);
                        }
                    }
                } else {
                    logStatus("PIN entry cancelled.");
                }

            } catch (IOException e) {
                String message = "NFC Communication Error: " + e.getMessage();
                logStatus(message);
                Log.e(TAG, message, e);
            } catch (SatocashNfcClient.SatocashException e) {
                String message = String.format("Satocash Card Error: %s (SW: 0x%04X)",
                        e.getMessage(), e.getSw());
                logStatus(message);
                Log.e(TAG, message);
            } catch (Exception e) {
                String message = "An unexpected error occurred: " + e.getMessage();
                logStatus(message);
                Log.e(TAG, message);
            } finally {
                try {
                    if (satocashClient != null) {
                        satocashClient.close();
                        Log.d(TAG, "NFC connection closed.");
                    }
                } catch (IOException e) {
                    Log.e(TAG, "Error closing NFC connection: " + e.getMessage());
                }
                logStatus("NFC interaction finished. Ready for next action.");
            }
        }).start();
    }

    private interface PinDialogCallback {
        void onPin(String pin);
    }
}

File: app/src/main/java/com/electricdreams/numo/TopUpActivity.kt
package com.electricdreams.numo

import android.app.PendingIntent
import android.content.Intent
import android.graphics.Color
import android.nfc.NfcAdapter
import android.nfc.Tag
import android.nfc.tech.IsoDep
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.text.InputType
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.app.NavUtils
import com.google.android.material.snackbar.Snackbar

class TopUpActivity : AppCompatActivity() {

    private lateinit var proofTokenEditText: EditText
    private lateinit var topUpSubmitButton: Button
    private var nfcDialog: AlertDialog? = null
    private var rescanDialog: AlertDialog? = null
    private var processingDialog: AlertDialog? = null
    private var nfcAdapter: NfcAdapter? = null
    private var satocashClient: SatocashNfcClient? = null
    private var satocashWallet: SatocashWallet? = null
    private var pendingProofToken: String? = null
    private var savedPin: String? = null
    private var waitingForRescan: Boolean = false
    private val mainHandler = Handler(Looper.getMainLooper())
    private lateinit var rootView: View

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setTheme(R.style.Theme_Numo)
        setContentView(R.layout.activity_top_up)

        // Set up back button
        findViewById<View?>(R.id.back_button)?.setOnClickListener {
            NavUtils.navigateUpFromSameTask(this)
        }

        // Set up the toolbar (hidden but kept for compatibility if needed)
        val toolbar: Toolbar? = findViewById(R.id.toolbar)
        toolbar?.let {
            setSupportActionBar(it)
            supportActionBar?.apply {
                setDisplayHomeAsUpEnabled(true)
                title = getString(R.string.top_up_title)
            }
        }

        rootView = findViewById(android.R.id.content)
        proofTokenEditText = findViewById(R.id.top_up_amount_edit_text)
        topUpSubmitButton = findViewById(R.id.top_up_submit_button)

        // Handle incoming share intent
        intent?.let { incomingIntent ->
            val action = incomingIntent.action
            val type = incomingIntent.type
            if (Intent.ACTION_SEND == action && type == "text/plain") {
                val sharedText = incomingIntent.getStringExtra(Intent.EXTRA_TEXT)
                if (!sharedText.isNullOrEmpty()) {
                    pendingProofToken = sharedText
                    proofTokenEditText.setText(sharedText)
                    showStatusMessage(
                        getString(R.string.top_up_status_token_ready),
                        success = true
                    )
                    showNfcDialog()
                }
            }
        }

        nfcAdapter = NfcAdapter.getDefaultAdapter(this)
        if (nfcAdapter == null) {
            showStatusMessage(
                getString(R.string.top_up_status_nfc_not_available),
                success = false
            )
            topUpSubmitButton.isEnabled = false
        }

        topUpSubmitButton.setOnClickListener {
            val proofToken = proofTokenEditText.text.toString()
            if (proofToken.isNotEmpty()) {
                pendingProofToken = proofToken
                showStatusMessage(
                    getString(R.string.top_up_status_tap_card),
                    success = true
                )
                showNfcDialog()
            } else {
                showStatusMessage(
                    getString(R.string.top_up_status_missing_token),
                    success = false
                )
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return if (item.itemId == android.R.id.home) {
            NavUtils.navigateUpFromSameTask(this)
            true
        } else {
            super.onOptionsItemSelected(item)
        }
    }

    private fun showStatusMessage(message: String, success: Boolean) {
        mainHandler.post {
            val snackbar = Snackbar.make(rootView, message, Snackbar.LENGTH_LONG)
            val snackbarView = snackbar.view
            val textView: TextView = snackbarView.findViewById(com.google.android.material.R.id.snackbar_text)
            textView.setTextColor(Color.WHITE)
            snackbarView.setBackgroundColor(
                if (success) Color.parseColor("#4CAF50") else Color.parseColor("#F44336")
            )
            snackbar.show()
        }
    }

    private fun showNfcDialog() {
        mainHandler.post {
            val builder = AlertDialog.Builder(this, R.style.Theme_Numo)
            val dialogView = layoutInflater.inflate(R.layout.dialog_nfc_modern, null)
            builder.setView(dialogView)

            builder.setCancelable(true)
            nfcDialog = builder.create().also { it.show() }
        }
    }

    private fun showRescanDialog() {
        mainHandler.post {
            val builder = AlertDialog.Builder(this, R.style.Theme_Numo)
            val dialogView = layoutInflater.inflate(R.layout.dialog_nfc_modern, null)
            builder.setView(dialogView)

            dialogView.findViewById<TextView>(R.id.nfc_amount_display)?.text =
                getString(R.string.top_up_dialog_ready_message)

            dialogView.findViewById<TextView>(R.id.nfc_dialog_title)?.text =
                getString(R.string.top_up_dialog_rescan_title)

            dialogView.findViewById<TextView>(R.id.nfc_hint_text)?.apply {
                text = getString(R.string.top_up_dialog_rescan_hint)
                visibility = View.VISIBLE
            }

            builder.setCancelable(true)
            builder.setOnCancelListener {
                savedPin = null
                waitingForRescan = false
            }

            rescanDialog = builder.create().also { it.show() }
        }
    }

    private fun showProcessingDialog() {
        mainHandler.post {
            val builder = AlertDialog.Builder(this, R.style.Theme_Numo)
            val dialogView = layoutInflater.inflate(R.layout.dialog_nfc_modern, null)
            builder.setView(dialogView)

            dialogView.findViewById<TextView>(R.id.nfc_dialog_title)?.text =
                getString(R.string.top_up_dialog_processing_title)
            dialogView.findViewById<TextView>(R.id.nfc_amount_display)?.text =
                getString(R.string.top_up_dialog_processing_message)

            builder.setCancelable(false)
            processingDialog = builder.create().also { it.show() }
        }
    }

    private fun showPinDialog(callback: (String?) -> Unit) {
        mainHandler.post {
            val builder = AlertDialog.Builder(this)
            builder.setTitle(R.string.dialog_title_enter_pin)

            val layout = LinearLayout(this).apply {
                orientation = LinearLayout.VERTICAL
                val density = resources.displayMetrics.density
                val padding = (50 * density).toInt()
                val paddingVertical = (20 * density).toInt()
                setPadding(padding, paddingVertical, padding, paddingVertical)
            }

            val input = EditText(this).apply {
                inputType = InputType.TYPE_CLASS_NUMBER or InputType.TYPE_NUMBER_VARIATION_PASSWORD
                hint = getString(R.string.dialog_pin_hint)
            }
            layout.addView(input)

            val keypadLayout = LinearLayout(this).apply {
                orientation = LinearLayout.VERTICAL
                layoutParams = LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
                )
            }

            val buttons = arrayOf(
                arrayOf("1", "2", "3"),
                arrayOf("4", "5", "6"),
                arrayOf("7", "8", "9"),
                arrayOf("", "0", "DEL")
            )

            for (row in buttons) {
                val rowLayout = LinearLayout(this).apply {
                    orientation = LinearLayout.HORIZONTAL
                    layoutParams = LinearLayout.LayoutParams(
                        LinearLayout.LayoutParams.MATCH_PARENT,
                        LinearLayout.LayoutParams.WRAP_CONTENT
                    ).apply { weight = 1.0f }
                }

                for (text in row) {
                    val button = Button(this).apply {
                        setText(text)
                        layoutParams = LinearLayout.LayoutParams(
                            0,
                            LinearLayout.LayoutParams.WRAP_CONTENT
                        ).apply { weight = 1.0f }
                        setOnClickListener {
                            when (text) {
                                "DEL" -> if (input.length() > 0) {
                                    input.text.delete(input.length() - 1, input.length())
                                }
                                "" -> Unit
                                else -> input.append(text)
                            }
                        }
                    }
                    rowLayout.addView(button)
                }

                keypadLayout.addView(rowLayout)
            }

            val buttonLayout = LinearLayout(this).apply {
                orientation = LinearLayout.HORIZONTAL
                layoutParams = LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
                ).apply { topMargin = (20 * resources.displayMetrics.density).toInt() }
            }

            val cancelButton = Button(this).apply {
                text = getString(R.string.common_cancel)
                layoutParams = LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1.0f).apply {
                    rightMargin = (8 * resources.displayMetrics.density).toInt()
                }
            }

            val okButton = Button(this).apply {
                text = getString(R.string.common_ok)
                layoutParams = LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1.0f).apply {
                    leftMargin = (8 * resources.displayMetrics.density).toInt()
                }
            }

            buttonLayout.addView(cancelButton)
            buttonLayout.addView(okButton)

            layout.addView(keypadLayout)
            layout.addView(buttonLayout)
            builder.setView(layout)

            val dialog = builder.create()

            cancelButton.setOnClickListener {
                dialog.cancel()
                callback(null)
            }

            okButton.setOnClickListener {
                val pin = input.text.toString()
                dialog.dismiss()
                callback(pin)
            }

            dialog.setOnCancelListener { callback(null) }

            dialog.show()
        }
    }

    override fun onResume() {
        super.onResume()
        nfcAdapter?.let { adapter ->
            val pendingIntent = PendingIntent.getActivity(
                this,
                0,
                Intent(this, javaClass).addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP),
                PendingIntent.FLAG_MUTABLE
            )
            val techList = arrayOf(arrayOf(IsoDep::class.java.name))
            adapter.enableForegroundDispatch(this, pendingIntent, null, techList)
        }
    }

    override fun onPause() {
        super.onPause()
        nfcAdapter?.disableForegroundDispatch(this)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        if (NfcAdapter.ACTION_TECH_DISCOVERED == intent.action) {
            val tag: Tag? = intent.getParcelableExtra(NfcAdapter.EXTRA_TAG)
            if (tag != null) {
                handleNfcImport(tag)
            }
        }
    }

    private fun handleNfcImport(tag: Tag) {
        val token = pendingProofToken
        if (token.isNullOrEmpty()) {
            showStatusMessage(
                getString(R.string.top_up_status_no_token_set),
                success = false
            )
            return
        }

        if (waitingForRescan && !savedPin.isNullOrEmpty()) {
            processImportWithSavedPin(tag)
            return
        }

        waitingForRescan = false

        Thread {
            var tempClient: SatocashNfcClient? = null
            try {
                tempClient = SatocashNfcClient(tag)
                tempClient.connect()
                Log.d(TAG, "Connected to NFC card")

                satocashClient = tempClient
                satocashWallet = SatocashWallet(satocashClient!!)

                satocashClient!!.selectApplet(SatocashNfcClient.SATOCASH_AID)
                Log.d(TAG, "Satocash Applet found and selected!")

                satocashClient!!.initSecureChannel()
                Log.d(TAG, "Secure Channel Initialized!")

                try {
                    val importedCount = satocashWallet!!.importProofsFromToken(token).join()
                    showStatusMessage(
                        getString(R.string.top_up_status_success_imported, importedCount),
                        success = true
                    )

                    mainHandler.post {
                        nfcDialog?.takeIf { it.isShowing }?.dismiss()
                        pendingProofToken = ""
                        proofTokenEditText.setText("")
                    }
                    return@Thread
                } catch (e: RuntimeException) {
                    val cause = e.cause
                    if (cause is SatocashNfcClient.SatocashException) {
                        val statusWord = cause.sw
                        Log.d(TAG, String.format("Status Word received: 0x%04X", statusWord))

                        if (statusWord == 0x9C06) {
                            Log.d(TAG, "PIN authentication needed")

                            try {
                                satocashClient?.let { client ->
                                    client.close()
                                    Log.d(TAG, "NFC connection closed before PIN entry")
                                    satocashClient = null
                                }
                            } catch (ioe: java.io.IOException) {
                                Log.e(TAG, "Error closing NFC connection before PIN entry: ${ioe.message}")
                            }

                            val pinFuture = java.util.concurrent.CompletableFuture<String?>()
                            mainHandler.post {
                                nfcDialog?.takeIf { it.isShowing }?.dismiss()
                                showPinDialog { pin -> pinFuture.complete(pin) }
                            }

                            val enteredPin = pinFuture.join()
                            if (!enteredPin.isNullOrEmpty()) {
                                savedPin = enteredPin
                                waitingForRescan = true
                                mainHandler.post { showRescanDialog() }
                            } else {
                                showStatusMessage(
                                    getString(R.string.top_up_status_operation_cancelled),
                                    success = false
                                )
                            }
                            return@Thread
                        } else {
                            showStatusMessage(
                                getString(R.string.top_up_status_card_error, cause.message ?: ""),
                                success = false
                            )
                        }
                    } else {
                        showStatusMessage("Error: ${e.message}", success = false)
                    }
                }
            } catch (e: java.io.IOException) {
                showStatusMessage(
                    getString(R.string.top_up_status_nfc_error, e.message ?: ""),
                    success = false
                )
            } catch (e: SatocashNfcClient.SatocashException) {
                showStatusMessage(
                    getString(R.string.top_up_status_card_error, e.message ?: ""),
                    success = false
                )
            } catch (e: Exception) {
                showStatusMessage(
                    getString(R.string.top_up_status_generic_error, e.message ?: ""),
                    success = false
                )
            } finally {
                try {
                    if (satocashClient != null && !waitingForRescan) {
                        satocashClient?.close()
                        Log.d(TAG, "NFC connection closed in finally block.")
                        satocashClient = null
                    }
                } catch (e: java.io.IOException) {
                    Log.e(TAG, "Error closing NFC connection: ${e.message}")
                }

                mainHandler.post {
                    nfcDialog?.takeIf { it.isShowing }?.dismiss()
                }
            }
        }.start()
    }

    private fun processImportWithSavedPin(tag: Tag) {
        val pin = savedPin
        if (pin.isNullOrEmpty()) {
            Log.e(TAG, "No saved PIN available for import")
            showStatusMessage(
                getString(R.string.top_up_status_no_saved_pin),
                success = false
            )
            return
        }

        Thread {
            var tempClient: SatocashNfcClient? = null
            try {
                rescanDialog?.takeIf { it.isShowing }?.let { dialog ->
                    mainHandler.post { dialog.dismiss() }
                }

                mainHandler.post { showProcessingDialog() }

                tempClient = SatocashNfcClient(tag)
                tempClient.connect()
                Log.d(TAG, "Connected to NFC card for PIN import")

                satocashClient = tempClient
                satocashWallet = SatocashWallet(satocashClient!!)

                satocashClient!!.selectApplet(SatocashNfcClient.SATOCASH_AID)
                Log.d(TAG, "Satocash Applet found and selected!")

                satocashClient!!.initSecureChannel()
                Log.d(TAG, "Secure Channel Initialized!")

                Log.d(TAG, "Authenticating with saved PIN...")
                val authenticated = satocashWallet!!.authenticatePIN(pin).join()

                if (authenticated) {
                    Log.d(TAG, "PIN Verified! Card Ready.")

                    val token = pendingProofToken ?: ""
                    val importedCount = satocashWallet!!.importProofsFromToken(token).join()

                    waitingForRescan = false
                    savedPin = null

                    showStatusMessage(
                        getString(R.string.top_up_status_success_imported, importedCount),
                        success = true
                    )

                    mainHandler.post {
                        pendingProofToken = ""
                        proofTokenEditText.setText("")
                    }
                } else {
                    val message = getString(R.string.top_up_status_pin_verification_failed)
                    Log.e(TAG, message)
                    waitingForRescan = false
                    savedPin = null
                    showStatusMessage(message, success = false)
                }
            } catch (re: RuntimeException) {
                val reCause = re.cause
                val message = if (reCause is SatocashNfcClient.SatocashException) {
                    // Keep SW formatting in the logged message, user-facing string is separate
                    getString(
                        R.string.top_up_status_pin_verification_failed_detail,
                        "${reCause.message} (SW: 0x%04X)".format(reCause.sw)
                    )
                } else {
                    getString(R.string.top_up_status_auth_failed, re.message ?: "")
                }
                Log.e(TAG, message)
                waitingForRescan = false
                savedPin = null
                showStatusMessage(message, success = false)
            } catch (e: Exception) {
                val message = getString(
                    R.string.top_up_status_unexpected_error,
                    e.message ?: ""
                )
                Log.e(TAG, message)
                waitingForRescan = false
                savedPin = null
                showStatusMessage(message, success = false)
            } finally {
                try {
                    satocashClient?.let { client ->
                        client.close()
                        Log.d(TAG, "NFC connection closed.")
                        satocashClient = null
                    }
                } catch (e: java.io.IOException) {
                    Log.e(TAG, "Error closing NFC connection: ${e.message}")
                }

                mainHandler.post {
                    processingDialog?.takeIf { it.isShowing }?.dismiss()
                }
            }
        }.start()
    }

    companion object {
        private const val TAG = "TopUpActivity"
    }
}
